src/core/qgsvectorfilewriter.cpp

```diff
@@ -2840,7 +2840,11 @@ gdal::ogr_feature_unique_ptr QgsVectorFileWriter::createFeature( const QgsFeatur
       }
       else // wkb type matches
       {
-        QByteArray wkb( geom.asWkb( QgsAbstractGeometry::FlagExportTrianglesAsPolygons ) );
+        QgsAbstractGeometry::WkbFlags wkbFlags = QgsAbstractGeometry::FlagExportTrianglesAsPolygons;
+        if ( mOgrDriverName == QLatin1String( "ESRI Shapefile" ) )
```

Comment:
for consistency I suspect the addZValue(0) and addMValue(0) at lines  2801 and 2803 should be modified

Reply:
> Do you mean using nan (like in the last commit I just added?).

yes, and passing the new flag to asWkb of line 2827